File: bench/callbacks/callbacks.py
# Functions as values: closures, composition, and list helpers that call back
# into a function for every element.
def compose(f, g):
    return lambda x: f(g(x))


double = lambda x: x * 2
increment = lambda x: x + 1
step = compose(increment, double)

xs = []
i = 0
while i < 2000:
    xs.append(i)
    i = i + 1


def map_(fn, values):
    out = []
    k = 0
    while k < len(values):
        out.append(fn(values[k]))
        k = k + 1
    return out


def filter_(fn, values):
    out = []
    k = 0
    while k < len(values):
        if fn(values[k]):
            out.append(values[k])
        k = k + 1
    return out


def reduce_(fn, values, initial):
    acc = initial
    k = 0
    while k < len(values):
        acc = fn(acc, values[k])
        k = k + 1
    return acc


total = 0
round_ = 0
while round_ < 1600:
    mapped = map_(step, xs)
    kept = filter_(lambda x: x % 3 == 0, mapped)
    total = total + reduce_(lambda acc, x: acc + x, kept, 0)
    round_ = round_ + 1
print(total)

File: bench/callbacks/callbacks.lua
-- Functions as values: closures, composition, and list helpers that call back
-- into a function for every element.
local function compose(f, g)
    return function(x) return f(g(x)) end
end

local double = function(x) return x * 2 end
local increment = function(x) return x + 1 end
local step = compose(increment, double)

local xs = {}
local i = 0
while i < 2000 do
    xs[#xs + 1] = i
    i = i + 1
end

local function map(fn, values)
    local out = {}
    local k = 1
    while k <= #values do
        out[k] = fn(values[k])
        k = k + 1
    end
    return out
end

local function filter(fn, values)
    local out = {}
    local k = 1
    while k <= #values do
        if fn(values[k]) then out[#out + 1] = values[k] end
        k = k + 1
    end
    return out
end

local function reduce(fn, values, initial)
    local acc = initial
    local k = 1
    while k <= #values do
        acc = fn(acc, values[k])
        k = k + 1
    end
    return acc
end

local total = 0
local round = 0
while round < 1600 do
    local mapped = map(step, xs)
    local kept = filter(function(x) return x % 3 == 0 end, mapped)
    total = total + reduce(function(acc, x) return acc + x end, kept, 0)
    round = round + 1
end
print(total)
